client/a8c-for-agencies/hooks/use-track-provisioning-sites.ts

```diff
@@ -0,0 +1,68 @@
+import { useState, useEffect, useCallback } from 'react';
+
+const PROVISIONING_SITE_CHANGE_EVENT = 'a4a-provisioning-site-change';
+const PROVISIONING_TIMEOUT = 300000; // 5 minutes in miliseconds
+
+type ProvisioningSite = {
+	id: number;
+	migration?: boolean;
+	development?: boolean;
+	ttl: number;
+};
+
+export default function useTrackProvisioningSites() {
+	const [ provisioningSites, setProvisioningSites ] = useState< ProvisioningSite[] >( [] );
+
+	const reload = useCallback( () => {
+		const storedSites = localStorage.getItem( 'provisioningSites' );
+		if ( storedSites ) {
+			const sites: ProvisioningSite[] = JSON.parse( storedSites );
+			// We remove the site from tracking if it is more than 5 minutes.
+			const activeSites = sites.filter( ( { ttl } ) => ttl >= new Date().getTime() );
+
+			setProvisioningSites( activeSites );
+			localStorage.setItem( 'provisioningSites', JSON.stringify( activeSites ) );
+		}
+	}, [] );
+
+	const trackSiteId = useCallback(
+		(
+			siteId: number,
+			{ migration, development }: { migration?: boolean; development?: boolean } = {}
+		) => {
+			const updatedSites = [
+				...provisioningSites,
+				{ id: siteId, migration, development, ttl: new Date().getTime() + PROVISIONING_TIMEOUT },
+			];
+
+			localStorage.setItem( 'provisioningSites', JSON.stringify( updatedSites ) );
+			window.dispatchEvent( new CustomEvent( PROVISIONING_SITE_CHANGE_EVENT ) );
+		},
+		[ provisioningSites ]
+	);
+
+	const untrackSiteId = useCallback(
+		( siteId: number ) => {
+			const updatedSites = provisioningSites.filter( ( { id } ) => id !== siteId );
+
+			localStorage.setItem( 'provisioningSites', JSON.stringify( updatedSites ) );
+			window.dispatchEvent( new CustomEvent( PROVISIONING_SITE_CHANGE_EVENT ) );
+		},
+		[ provisioningSites ]
+	);
+
+	useEffect( () => {
+		window.addEventListener( PROVISIONING_SITE_CHANGE_EVENT, reload );
```

Comment:
Very neat solution! 👏  I am just wondering why is there a need to trigger the state changes via window event listeners? Why don't we call the `reload` callback directly when needed? 🤔

Reply:
We are currently using the same hook in two separate components (the header and the Dashboard component).

However, these components do not share the same state. When we trigger provisioning via the 'Create new site' button under the Header component, the Dashboard component, where the banner is located, is not aware of the change.

To resolve this issue, we have created a custom event that all-consuming components can listen to and be notified of the event.

Another potential solution to this problem could be using the Context API, but this would require us to provide the context at the root of the page, which might not always be ideal for other use cases.

Reply:
Got it. Thanks for the explanation. 🙂